### DISCUSSION TITLE
Difference between having an empty `paths` array or not in `getStaticProps()`

### DISCUSSION THREAD

**Discussion by Danetag** [vercel/next.js#16045] [Help] [answered]
Hey team!

First of all, thanks for the great work you're all doing!

A few little questions. I'd love to generate as many static pages as possible for an e-commerce website I'm working on.

- `revalidate`: Is the page rebuilt _once_ x seconds after a user visits the page, or is is rebuild _every_ x seconds?
- If it's possible to (re)build on runtime a page statically, what would be the advantages of providing a `paths` array? What would a bot see the first time it visits a page not rendered yet? How would that page be removed if it's not supposed to exist anymore?

Thank you!

**Comment by santiagopoli** [comment, marked as answer]
* Revalidate rebuilds the page upon the first request after the provided threshold. It is important to note that this first request will still receive the cached version, as the revalidation process runs in the background (but subsequent requests will get the new one). If your threshold is 10 seconds but you don't receive a new request after lets say 60 seconds, your page will be rebuilt at those 60 seconds (I hope I'm making myself clear)

* The advantage would be having pages rendered right after a deploy and not having to wait for those pages to build for the first request. What will a bot see? It depends on the value of the `fallback` field in your getStaticPaths. If you set `fallback: true`, you can check `router.isFallback` in your component to show a Loader. On the other hand, you can set `fallback: 'unstable_blocking'` (it is an experimental feature) and block the first request until the page finishes to build. There is also `fallback: false` which will return a 404 but I personally don't recommend it.

Relevant docs: https://nextjs.org/docs/basic-features/data-fetching#the-fallback-key-required

P.S. Even if you show a Loader while the page is being built (using `fallback: true`), Google crawlers will still be able to index your page (they support JavaScript now). I'm not sure about other crawlers though.

**Reply by Danetag** [reply]
Thanks for the quick answer! If I use `paths:[]` and `fallback: true|unstable_blocking`, what happens if an existing page needs to be deleted?

**Reply by santiagopoli** [reply]
If by deleting a page you mean deleting a file (like page1.js), you won't have a problem since caches are automatically purged on every deploy (well, tecnically they are not purged, but linked to the previous deployment). This assumes you are using Vercel as a platform.

**Reply by Danetag** [reply]
Oh sorry, I was referring in the context of a dynamic page. 

For instance, let's say we have `brands/[brand].tsx`, if a brand page exists (`/brands/a-brand`) but later on needs to be deleted because that brand is not supported anymore. 

What would happen if we use `getStaticProps()` with `revalidate: 1` + `getStaticPaths()`  with `paths:[]` and `fallback: true|unstable_blocking` in this case?

**Reply by santiagopoli** [reply]
Suppose this scenario:

```
export async function getStaticPaths() {
  return {
    paths: [],
    fallback: true,
  }
}
```
```
export async function getStaticProps({ params }) {
  const { brandId } = params
  const brand = await getBrand(brandId)

  if(brand.notSupported) {
    return {
      props: {},
      revalidate: 1,
    }
  }

  return {
    props: { brand },
    revalidate: 1,
  }
}
```
```
export default function BrandPage({ brand }) {
  const router = useRouter()
  if (router.isFallback) return <Loading/>
  if(!brand) return <BrandNotFound/>

  return <Brand brand={brand}/>
}
```

In this case, if you disable a brand, the first requests (after 1 second since the last build) will still see it, but subsequent requests will see the `<BrandNotFound/>` component.

If you use `fallback: 'unstable_blocking'` you can effectively remove these lines:

```
const router = useRouter()
if (router.isFallback) return <Loading/>
```

**Reply by Danetag** [reply]
Awesome! thank you so much for those examples 🙏

**Reply by Cassianokunsch** [reply]
Hello everyone, I have a question about it. In this case the files brandId.html and brandId.json (where brandid is a key of the brand, for each brand), still exists on server?
